samples/custom_dispatch/cuda/kernels/ukernel.cu

```diff
@@ -0,0 +1,11 @@
+extern "C" __device__ void simple_mul_workgroup(float *lhs, size_t lhs_offset,
```

Comment:
C++ templates are necessary to support multiple data types, as writing multiple microkernels in C is not scalable. For the `topk` operation, we require at least 10 microkernels, and for `linalg.matmul`, around 40-50 microkernels are needed.

Regarding the LowerUKernelsToCalls Pass, can we call name mangling? I guess it is done in clang, maybe it's not trivial :/

We could explore the option of using a microkernel generator. I have implemented a [microkernel generator](https://github.com/grypp/iree/blob/microkernel/compiler/src/iree/compiler/Codegen/Microkernels/CUDA/uCUDAKernelGenerator.cpp) 
, which you can find here on issue #12026.

Reply:
As a third way, we can implement a single microkernel with a GIGA switch-case. For example, the first topk kernel is below.

```
template <typename T, typename K, unsigned int CTA_SIZE>
__global__ __launch_bounds__(CTA_SIZE) void topkPartial(
    T *__restrict__ inputs,            // input
    T *__restrict__ outPartialVals,    // output of partial topk  values
    K *__restrict__ outPartialIndices, // output of partial topk
                                       // indices
    int kValue,                        // K value
    K rank,                            // rank tensor<> (1D, 2D, 3D, ...)
    K elemsIn1dim, K elemsIn2dim) ...
```

We can implement a single microkernel with a giga switch case. In this case, we should pass the data types and all the other things as function parameter. 

```
enum UkernelType{
	float,
	int,
	...
};

__device__ extern "C" void GIGA_topkPartial(
	UkernelType TypeK, UkernelType TypeT, int CTA_SIZE, 
    void *__restrict__ inputs,        
    void *__restrict__ outPartialVals,   
    void *__restrict__ outPartialIndices,
                                      
    int kValue,                       
    void rank,                           
    void elemsIn1dim, void elemsIn2dim) {
    
    switch{
	case(...):
	 topkPartial<float,float,128>(inputs, outPartialVals, outPartialIndices, kValue, rank, elemsIn1dim, elemsIn2dim);
	case(...):
	 topkPartial<float,int,128>(inputs, outPartialVals, outPartialIndices, kValue, rank, elemsIn1dim, elemsIn2dim);
	case(...):
	 topkPartial<float,float,256>(inputs, outPartialVals, outPartialIndices, kValue, rank, elemsIn1dim, elemsIn2dim);
	/// many more kernel call
	default:
	error
    };
}
```

I remember we talked about that in the past. I am reluctant of doing this as it might causes register size bloat.

Reply:
> C++ templates are necessary to support multiple data types, as writing multiple microkernels in C is not scalable. For the `topk` operation, we require at least 10 microkernels, and for `linalg.matmul`, around 40-50 microkernels are needed.

I would highly highly recommend using reflection like NVIDIA libdevice does. We already have this working on CPU side. We can pass a data structure that has the processor information that you can use to navigate the microkernel selection. All that gets constant folded (this already works).

> 
> 1. Can we call name mangling to call C++ `LowerUKernelsToCalls` Pass? I guess it is done in clang, maybe it's not trivial to do it IREE.

Mangling is actually not standard across compilers... The only robust way of doing this is through extern "C" functions (this has nothing to do with IREE, its just about LLVM linking).

> 2. As second solution, we can use macros in C to mimic templates and we can implement a generator. I have implemented on issue [[WIP] Microkernels for CUDA target #12026](https://github.com/openxla/iree/pull/12026).
>    [C microkernel mimics C++ templates](https://github.com/grypp/iree/blob/microkernel/compiler/src/iree/compiler/Codegen/Microkernels/CUDA/uCUDAGemmTemplate.cuh?rgh-link-date=2023-06-26T10%3A52%3A46Z#L15-L24)
>    [Generator](https://github.com/grypp/iree/blob/microkernel/compiler/src/iree/compiler/Codegen/Microkernels/CUDA/uCUDAKernelGenerator.cpp?rgh-link-date=2023-06-26T10%3A52%3A46Z#L82-L84)

We should not be using any of these... we should just be using reflection.

> I remember we talked about that in the past. The main disadvantage of this approach is that the template parameters are not visible at compile time of microkernel. This may cause bad performance.

That is not true. This all gets resolved at compile time (we already have this working on the CPU side).